### DISCUSSION TITLE
Requests for commonly used rust/cargo

### DISCUSSION THREAD

**Discussion by panda2134** [verus-lang/verus#1129] [Feature requests]
This discussion collects a list of rust language features frequently used in coding, yet unsupported by verus.

**Comment by panda2134** [comment]
### Verification on explicit type conversion is not supported

In rust language, `From` and `Into` traits are used to implement type conversion. Usually, the user only implements the `From` trait, and the rust compiler calls it whenever a `.into()` is called.

However, the verus compiler cannot correctly handle these two traits. A minimal example is provided below:

```rust
use vstd::prelude::*;
use vstd::string::new_strlit;

verus! {
    #[repr(u8)]
    #[derive(Debug, PartialEq, Eq)]
    pub enum FileType {
        Dir = 1,
        File = 2,
    }
    
    impl From<u32> for FileType {
        #[verifier(external_body)]
        fn from(val: u32) -> Self {
            match val {
                1 => FileType::Dir,
                2 => FileType::File,
                _ => panic!("Invalid file type"),
            }
        }
    }
    
    impl From<FileType> for u32 {
        #[verifier(external_body)]
        fn from(val: FileType) -> Self {
            match val {
                FileType::Dir => 1,
                FileType::File => 2,
            }
        }
    }

    fn main() {
        let val = FileType::Dir;
        let t: u32 = val.into();
        assert(t >= 1 && t <= 2);
    }
}
```

When one attempts to verify the file with verus, the following error is generated, indicating that verus does not correctly handle `val.into()`:

```
error: `core::convert::impl&%3::into` is not supported (note: you may be able to add a Verus specification to this function with the `external_fn_specification` attribute) (note: the vstd library provides some specification for the Rust std library, but it is currently limited)
  --> 1.rs:35:22
   |
35 |         let t: u32 = val.into();
   |                      ^^^^^^^^^^
```

**Reply by utaal** [reply]
From triage:

You can do it for whatever type you want https://doc.rust-lang.org/std/convert/trait.From.html

? desugars into something that has a trait bound that interacts with FromResidual.

* [ ] proper spec for From and remove hack for ?.

One can add their own external spec.

* [ ] document that process.

**Reply by tjhance** [reply]
I checked and it turns out From/Into are already supported in vstd, with `into` given the following spec:

```
pub assume_specification<T, U: From<T>>[ T::into ](a: T) -> (ret: U)
    ensures
        call_ensures(U::from, (a,), ret),
```

**Reply by utaal** [reply]
Shall I mark this resolved?

**Comment by panda2134** [comment]
### Vector literal is not supported

It is common in rust code that a vector is define with macros. For instance,  `vec![3, 4, 5]` defines a vector with three i32 elements. This is not properly supported with verus. Consider the following code:

```rust
use vstd::prelude::*;
use vstd::std_specs::core;

verus! {

    fn main() {
        let v = vec![1, 2, 3];
    }
}
```

For this file, verus halts with the following error:

```
error: The verifier does not yet support the following Rust feature:
unsizing operation from `std::boxed::Box<[i32; 3]>` to `std::boxed::Box<[i32]>`
```

**Reply by utaal** [reply]
Note: `vec![1, 2, 3]` desugars into `Box::new([1, 2, 3]).to_vec()` (or something like that).

**Reply by tjhance** [reply]
`vec!` is now supported

**Comment by panda2134** [comment]
### Integration with cargo

It would be better if we can add some documentation on using verus with cargo, so cargo can add all build arguments about dependencies to the verus call. This seems to be doable with [[custom toolchains](https://rust-lang.github.io/rustup/concepts/toolchains.html#custom-toolchains)](https://rust-lang.github.io/rustup/concepts/toolchains.html#custom-toolchains), but we need better documentation about this.

If we look into [[source code of rustup](https://github.com/rust-lang/rustup/blob/2a5a69e0914ff419554d684ca71eb1d72c72bcb3/src/cli/rustup_mode.rs#L1240)](https://github.com/rust-lang/rustup/blob/2a5a69e0914ff419554d684ca71eb1d72c72bcb3/src/cli/rustup_mode.rs#L1240), we can note that a valid rust toolchain only requires the `lib/` directory, and the rustc compiler `bin/rustc`. Maybe we can make a toolchain that includes verus as the "rustc". By switching to this toolchain, the user can then use verus to verify and compile a cargo project.

**Comment by utaal** [comment]
### Basic vstd functions for (debug) printing in verified code

`println!` and friends are not supported in verified code yet. We can provide simpler, trusted alternatives for those for now.
As requested by @panda2134.

**Reply by utaal** [reply]
@jaybosamiya: something that's easier than println

It's supposed to be transparent. Encode the body, ignore the dbg!.

Add Debug to the ignored traits.

@Chris-Hawblitzel: find all the things we want Verus to ignore.

@jaybosamiya-ms: Specifically, dbg! (https://doc.rust-lang.org/std/macro.dbg.html)

@utaal https://github.com/verus-lang/verus/discussions/1129#discussioncomment-

@Chris-Hawblitzel treat the Debug impl as external, and enforce it to be called from the dbg! macro.

Add an attribute "this comes from the dbg! macro".
